features/mbedtls/targets/TARGET_NUVOTON/TARGET_M480/des/des_alt.c

```diff
@@ -342,66 +361,71 @@ static int mbedtls_des_docrypt(uint16_t keyopt, uint8_t key[3][MBEDTLS_DES_KEY_S
     *tdes_key ++ = val;
     val = nu_get32_be(key[2] + 4);
     *tdes_key ++ = val;
-    
+
     uint32_t rmn = length;
     const unsigned char *in_pos = input;
     unsigned char *out_pos = output;
-    
-    while (rmn) {
+
+    while (rmn > 0) {
         uint32_t data_len = (rmn <= MAXSIZE_DMABUF) ? rmn : MAXSIZE_DMABUF;
-        
+
         uint32_t ivh, ivl;
         ivh = nu_get32_be(iv);
         ivl = nu_get32_be(iv + 4);
         TDES_SetInitVect(0, ivh, ivl);
-    
+
         memcpy(dmabuf_in, in_pos, data_len);
-        
+
         TDES_SetDMATransfer(0, (uint32_t) dmabuf_in, (uint32_t) dmabuf_out, data_len);
-        
+
         // Start enc/dec.
         // NOTE: Don't call driver function TDES_Start in BSP because it will override TDES_CTL[3KEYS] setting.
         CRPT->TDES_CTL |= CRPT_TDES_CTL_START_Msk | (CRYPTO_DMA_ONE_SHOT << CRPT_TDES_CTL_DMALAST_Pos);
```

Comment:
I am not sure the idiom used here is valid; there doesn't seem to be any countermeasure against memory-reordering  -- neither on the compiler- nor on the processor-level -- that would ensure that the cryptoaccelerator sees the updated contents of `dmabuf_in` when `TDES_CTL` is written. I'm not an expert on this, though - @0xc0170 ?

Reply:
@hanno-arm Maybe you misunderstood it. I just wanted to prevent `CRPT->TDES_CTL` from overridden by BSP driver's internal multiple channel support. Anyway, I reworked it by calling `TDES_Start` in https://github.com/ARMmbed/mbed-os/pull/4925/commits/5f229a3d33096787ad8a312bb561159a432fb7c0.

Reply:
@ccli8 I didn't mean to refer to the comment in the code, but to the entire path that (1) moves data to the DMA input buffer, (2) sets the start flag, (3) waits for the result flag, (4) fetches data from the DMA output buffer. It appears susceptible for memory reordering:
1. For example, as far as I see there is no barrier that would forbid the compiler to move reading the DMA output buffer before the loop polling on `CRPT->TDES_STS`. 
2. There are no processor-level barriers, but this might not be necessary here because the target is an M4 - I haven't checked in detail.

Reply:
@hanno-arm By referring to **multi-master systems** section in [ARM Cortex-M Programming Guide to Memory Barrier Instructions](https://static.docs.arm.com/dai0321/a/DAI0321A_programming_guide_memory_barriers_for_m_profile.pdf), I added memory barrier in https://github.com/ARMmbed/mbed-os/pull/4925/commits/6b7190b00729c3dbf3203019e934b59e41a8f401 for safe DMA transfer.